util/topgen.py

```diff
@@ -731,29 +731,33 @@ def render_template(template_path: str, rendered_path: Path, **other_info):
         with rendered_path.open(mode="w", encoding="UTF-8") as fout:
             fout.write(template_contents)
 
-    # The Rust file needs some complex information, so we initialize this
-    # object to store it.
-    rs_helper = TopGenRust(completecfg, name_to_block, version_stamp)
-
-    rust_files = [("toplevel_mod.rs.tpl", "mod.rs"),
-                  ("toplevel.rs.tpl", f"top_{topname}.rs")]
-
     # Create Rust output directory
     rsformat_dir = out_path / "sw/autogen/chip/"
     rsformat_dir.mkdir(parents=True, exist_ok=True)
 
     # Generating Rust device description for external sw usage
-    for (template, source) in rust_files:
-        render_template(topgen_template_path / template,
-                        rsformat_dir / source,
+    render_template(topgen_template_path / 'toplevel_mod.rs.tpl',
+                    rsformat_dir / 'mod.rs')
+
+    for addr_space in completecfg['addr_spaces']:
+        addr_space_suffix = lib.get_addr_space_suffix(addr_space)
+
+        # The Rust file needs some complex information, so we initialize this
+        # object to store it.
+        rs_helper = TopGenRust(completecfg, name_to_block, version_stamp)
+
+        render_template(topgen_template_path / 'toplevel.rs.tpl',
+                        rsformat_dir / f"top_{topname}{addr_space_suffix}.rs",
+                        addr_space=addr_space['name'],
                         helper=rs_helper)
 
-    # Generate Rust host-side files
-    rsformat_dir = src_tree_top / "sw/host/opentitanlib/src/chip/autogen"
-    rsformat_dir.mkdir(parents=True, exist_ok=True)
-    render_template(topgen_template_path / "host_toplevel.rs.tpl",
-                    rsformat_dir / f"{topname}.rs",
-                    helper=rs_helper)
+        # Generate Rust host-side files
+        rsformat_dir = src_tree_top / 'sw/host/opentitanlib/src/chip/autogen'
```

Comment:
Note that this change of `rsformat_dir` inside the address space loop also affects the generation of the following device-side files which is wrong. I've now filed #25953 to fix this and to unblock CI again.

Reply:
Nice catch! I didn't pay close enough attention to the rust outputs, apparently. That code could be cleaned up a little more...

Reply:
Don't worry, it's working now :-)